fedimint-client/src/module/init/recovery.rs

```diff
@@ -0,0 +1,417 @@
+            // of `recover` function not yet persisted in the database. So
+            // it's possible that `recovery` would be called again on an
+            // already finalized state. Because of this we need to store some
+            // finalization marker in the same dbtx as the finalization itself, detect this
+            // here and exit early.
+            warn!("Previously finalized, exiting");
+            return Ok(());
+        }
+        let current_session_count = client_ctx.global_api().session_count().await?;
+
+        let (mut state, mut common_state) =
+            // TODO: if load fails, we could ... start from scratch? at least being able to force this behavior might be useful
+            if let Some((state, common_state)) = Recovery::load_dbtx(&mut db.begin_transaction_nc().await, self).await {
+                (state, common_state)
+            } else {
+                let (state, start_session) = Recovery::new(self, snapshot).await?;
+                (state,
+                RecoveryFromHistoryCommon {
+                    start_session,
+                    next_session: start_session,
+                    end_session: current_session_count + 1,
+                })
+            };
+
+        let mut block_stream = fetch_block_stream(
+            self.api().clone(),
+            client_ctx.decoders(),
+            common_state.next_session..common_state.end_session,
+        );
+        let client_ctx = self.context();
+
+        while common_state.next_session < common_state.end_session {
+            make_progress(
+                &client_ctx,
+                &mut common_state,
+                &mut state,
+                &mut block_stream,
+            )
+            .await?;
+
+            let mut dbtx = db.begin_transaction().await;
+            state.store_dbtx(&mut dbtx.to_ref_nc(), &common_state).await;
+            dbtx.commit_tx().await;
+
+            self.update_recovery_progress(RecoveryProgress {
+                complete: (common_state.next_session - common_state.start_session)
+                    .try_into()
+                    .unwrap_or(u32::MAX),
+                total: (common_state.end_session - common_state.start_session)
+                    .try_into()
+                    .unwrap_or(u32::MAX),
+            })
+            .await
+        }
+
+        debug!(
+            target: LOG_CLIENT_RECOVERY,
+            ?state,
+            "Finalizing restore"
+        );
+
+        client_ctx
+            .clone()
+            .module_autocommit_2(
+                move |dbtx, _| {
+                    let state = state.clone();
+                    {
+                        Box::pin(async move {
+                            state.delete_dbtx(&mut dbtx.module_dbtx()).await;
+                            state.finalize_dbtx(dbtx).await?;
+                            Recovery::store_finalized(&mut dbtx.module_dbtx(), true).await;
```

Comment:
Further down you write that finalization can happen without recovery being terminated/removed from DB? I can't see how.

Reply:
I missed this problem in all previous implementations, as it's easy to overlook.

It can't happen anymore because of this `state.finalize_dbtx` and the check where this comment was explaining why it's necessary.

Example sequence how this happens (if `finalize_dbtx` didn't exist):

0. module progress is complete and saved to the db
1. `dbtx` with finalization commits, completing recovery on the client module side
2. client crashes/gets terminated
3. client starts again
4. client never observed/persisted that the module finished recovery, so calls module recovery again
5. module loads progress and sees it's complete, moves on to finalization again
6. module runs finalization again and probably fails because it's actually not idempotent and doesn't expect the already existing state.